File: blockchain-in-60-lines-of-ts/block.ts
import * as crypto from 'crypto';

const SHA256 = (message: string) => crypto.createHash('sha256').update(message).digest('hex');

export class Block {
  data: Object;
  hash: string;
  nonce: number;
  prevHash: string;
  timestamp: string;

  constructor(timestamp: string = '', data: Object = []) {
    this.data = data;
    this.hash = this.getHash();
    this.nonce = 0;
    this.prevHash = '';
    this.timestamp = timestamp;
  }

  public getHash(): string {
    return SHA256(this.prevHash + this.timestamp + JSON.stringify(this.data) + this.nonce);
  }

  public mine(difficulty: number) {
    // Basically, it loops until our hash starts with
    // the string 0...000 with length of <difficulty>.
    while (!this.hash.startsWith(Array(difficulty + 1).join('0'))) {
      this.nonce++;
      // Update our new hash with the new nonce value.
      this.hash = this.getHash();
    }
  }
}

File: PoW-blockchain-python/block.py
import json
from hashlib import sha256


class Block:
    def __init__(self, data=[], timestamp="") -> None:
        self.data = data
        self.nonce = 0
        self.prev_hash = ""
        self.timestamp = timestamp
        self.hash = self.getHash()

    def getHash(self):
        return sha256(
            (f"{self.nonce}{self.prev_hash}{self.timestamp}{json.dumps(self.data)}").encode("utf-8")
        ).hexdigest()

    def mine(self, difficulty):
        while not self.hash.startswith((difficulty + 1) * "0"):
            self.nonce += 1
            self.hash = self.getHash()
